### ISSUE TITLE
`.server` -- syntax shortcut for setting an executor?

### ISSUE THREAD

**Issue by liamhuber** [pyiron/pyiron_workflow#475]
#456 is two issues, so I'm going to break out the server stuff here.

**Comment by liamhuber** [comment]
@JNmpi wrote:

### Some ideas/thoughts to control execution

#### Direct approach
```python
    wf.nodes = wf.create.my_path.MyNode(a=1, b=2).server(cores=4).iter(element=['Al', 'Fe', 'Pb']).server(workers=3, queue='test')
```

Run 3 MyNode jobs simultaneously on 4 cores each, i.e., the queue test is assumed/told to provide 12 cores. This notation can be easily extended to include additional control variables (e.g. for controling the log level, gpu vs cpu), etc.    

#### Indirect approach via workflow

```python
    wf = Workflow('my_workflow').server(cores=12, queue='test')

    # run interactively (no submission to queue)
    wf.interactive_node = wf.create.my_path.MyNode(a=1, b=2) 

    # run by submitting to queue (note: the number of workers is not needed, is obtained from 12 cores/ 4 cores per node)
    wf.expensive_nodes = wf.create.my_path.MyNode(a=1, b=2).server(cores=4, to_queue=True).iter(element=['Al', 'Fe', 'Pb'])

    wf.run()
```

Alternative formulation (both could be supported)

```python
    wf = Workflow('my_workflow')

    # run interactively (no submission to queue)
    wf.interactive_node = wf.create.my_path.MyNode(a=1, b=2) 

    # run by submitting to queue (note: the number of workers is not needed, it is obtained from 12 cores/ 4 cores per node)
    wf.expensive_nodes = wf.create.my_path.MyNode(a=1, b=2).server(cores=4, to_queue=True).iter(element=['Al', 'Fe', 'Pb'])

    wf.server(cores=12, queue='test').run()
```

### Further extensions of the server 

```python
    server(
        hash_node = None, # take from node definition (default is False)
        log_level = wf.definitions.log_level.default # .debug, .production etc. via autocompletion
        
    )
```

Alternatively to 'server' we may consider adding additional objects that control hashing, output etc. to prevent overloading 'server' with input arguments


```python

```

**Comment by liamhuber** [comment]
@liamhuber wrote:

Currently we can set `some_node.executor` (or `some_for_node.body_node_executor` as convenience for for-loops) to _any_ `concurrent.futures.Executor` and it will just work. This flexibility is what let's us submit to `concurrent.futures.ProcessPoolExecutor` just as easily as we submit to `executorlib.Executor` [running on slurm](https://github.com/pyiron/pyiron_workflow/issues/453). So, I don't want to give up that simplicity and power, but could see `.server` as a shortcut to access the same power.

The pro to your suggestions here is that the syntax is quite tight. The con is creating a much tighter coupling between the `pyiron_workflow` code and some particular executor implementation, i.e. the ongoing maintenance work to ensure that `pyiron_workflow` stays synchronized with whatever it is that we're binding to. I agree that the `.server()` stuff is slightly tighter than `.executor =`, I'm just not 100% convinced it's enough snappier to be worth the extra maintenance/providing a second interface to executors [("There should be one-- and preferably only one --obvious way to do it.")](https://peps.python.org/pep-0020/)

**Comment by liamhuber** [comment]
@samwaseda wrote:

On the other hand, [in contrast to `.iter` stuff,] I would be more concerned about the idea with `server`, because the server information for me is a node attribute, i.e. I can imagine people writing something like:

```python
wf.something = my_path.Something("foo")
wf.something.server(…)
```

In this case not only `something` does not get the server information, but the node does not exist anymore.

**Comment by liamhuber** [comment]
Oh, I thought that example would run fine, @samwaseda. In my head:

```python

def Node.server(self, **a_bunch_of_kwargs):
    self.executor = self._build_an_executor_from_server_kwargs(**a_bunch_of_kwargs)
    return self
```

Of course `_build_an_executor_from_server_kwargs` could be in-lined in that function, I'm just pseudocoding. 

Then in

```python
wf.something = my_path.Something("foo")
wf.something.server(…)
```

The first line instantiates a `Something` node, and the second node internally applies an executor to it and returns the node. The node return simply does nothing here, but is useful in @JNmpi's other example where we chain it together. 

Since it doesn't matter if we apply the `executor` object earlier or later, then (in addition to `.iter` returning a node instance, as discussed in #456), returning `self` makes the server call transitive such that this:

```python
wf.create.my_path.MyNode(a=1, b=2).server(cores=4, to_queue=True).iter(element=['Al', 'Fe', 'Pb'])
```

and this

```python
wf.create.my_path.MyNode(a=1, b=2).iter(element=['Al', 'Fe', 'Pb']).server(cores=4, to_queue=True)
```

are equivalent.

Where I see things breaking down is in the pseudocode examples where two separate `.server` calls are made to somehow nest together. Right now `Node` parallelizes using any `concurrent.futures.Executor`-compliant executor, namely `executorlib.Executor`, where it is a bit verbose but otherwise totally straightforward to run on SLURM. When I try nesting simple `concurrent.futures.ProcessPoolExecutors` it doesn't work, and afaik [nesting executors is not working with `executorlib` yet either](https://github.com/pyiron/executorlib/issues/204). So we might be able to get this sort of two-shot behaviour in the future, but in the meantime we can make `.server(...)` a shortcut for one-shot setting the executor we want.

Also, just like we have `For.body_node_executor` that is a shortcut to setting `.executor` in all children of for-loop node, we could analogously have `.body_node_server(...)`. I.e. instead of 

```python
wf.nodes = wf.create.my_path.MyNode(a=1, b=2).server(cores=4).iter(element=['Al', 'Fe', 'Pb']).server(workers=3, queue='test')
```

We'd have something like

```python
wf.nodes = wf.create.my_path.MyNode(a=1, b=2).iter(element=['Al', 'Fe', 'Pb']).body_node_server(cores=4, workers=3, queue='test')
```

Not that I have any fundamental objection to the former, just that I don't see us getting it technically implemented any time soon.

Overall after letting it percolate for a couple days, my positions is the same: I don't object to this idea, but I think we should play around with running stuff on SLURM _at all_ before jumping into an abstraction. In addition, whatever we do to couple to whatever is giving us cluster-capable executors (presumably `executorlib`) should be the thinnest interface we can manage in order to minimize upkeep costs -- even to the point that it may be worth losing the tab-completion of `.server(...)` in exchange for making the server abstractions somewhere else entirely and sticking with `.executor = some_lib.Server(...)`.

**Comment by samwaseda** [comment]
> Oh, I thought that example would run fine, @samwaseda. In my head:
> 
> ```python
> def Node.server(self, **a_bunch_of_kwargs):
>     self.executor = self._build_an_executor_from_server_kwargs(**a_bunch_of_kwargs)
>     return self
> ```

Yes, that's a possibility, which I also thought about, and technically it's definitely super easy. My problem is that it mixes return functions and in-place functions, which would be super dangerous if we implement `node.iter()`. So before implementing `iter`, `server` etc., I think we should decide whether functions applied to nodes *always* return a node, or are *always* in-place, but not both and also not mixed.

**Comment by liamhuber** [comment]
Oh, that is extreme. We can certainly not have it that node methods always return a node -- e.g. `disconnect()` returns what got disconnected, which makes sense and is super helpful. Further, node inherits from other classes, and is inherited from by yet more classes -- it would be tough to consistently demand that every class in the inheritance tree always returns `self` from every public method.

The alternative extreme that methods _never_ return node -- or at least that we don't set up the syntax to actively encourage users to make in-place modifications -- conflicts directly with the last two ideas that he expressed very strong feelings about. I have some concern about each (the orphaning edge case, and extra maintenance to keep the server interface synced with the underlying executor), but also find both approaches visually appealing.

So always and never returning a node are neither feasible. I 100% see where you're coming from with concern about user confusion mixing and matching in-place and explicit approaches. Pythonically, I think we should avoid such in-place approaches because "explicit is better than implicit", "there should be one, and preferably only one way to do things", and "complicated is better than complex". But I'm also not set against the in-place wrappers because being able to do it by tab-completion is pyironic.

**Comment by samwaseda** [comment]
ok my comment was mainly with reference to `iter` but it's true that there are other common methods like `pop` which changes the object and returns something. But for me that would basically imply that `iter` as Jörg and I favoured should not be implemented in the way that we proposed. As a matter of fact, maybe it's even the right step to remove `iter()` and `zip()` because either they do something that differs from the rest of the workflow nodes (i.e. it doesn't delay) or it would be dangerous, as we discussed.

I guess `server` is safe enough to be implemented in the way that you suggested above, i.e.:

```python
def Node.server(self, **a_bunch_of_kwargs):
    self.executor = self._build_an_executor_from_server_kwargs(**a_bunch_of_kwargs)
    return self
```

**Comment by liamhuber** [comment]
Pre-`flowrep` issue
